### PR TITLE
feat(frontend): support `booleq` and `boolne` function

### PR DESCRIPTION
**Pull request by likzn**
## What's changed and what's your intention?
Support `booleq` and `boolne`. 
But it is not implemention totaly. Such as below, it will error. I think we can open an another PR to support them together for these functions(`IsTrue`,`And`,`Booleq` and so on). And a naive thought,  make rewrite when bind them.
```
SELECT booleq('true','true');
SELECT 1 where 'true' is true;
SELECT 1 where 'true' and 'true';
```

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
#2534

### CODE REVIEW

**Comment by likzn** [comment]
@TennyZhuang PTAL~

**Comment by neverchanje** [comment]
TBH I really didn't find a function definition named "booleq" in postgres's official document. Not even in materialize's doc: https://materialize.com/docs/sql/functions/. If this function does not exist in pg, we'd better don't implement it.

**Comment by likzn** [comment]
I find something. And i runed it in postgresql, it can work.

https://github.com/postgres/postgres/blob/8d061acd12af26551b607299aa204c946d2b6ba2/src/backend/utils/adt/bool.c#L229-L237

**Comment by TennyZhuang** [comment]
> TBH I really didn't find a function definition named "booleq" in postgres's official document. Not even in materialize's doc: https://materialize.com/docs/sql/functions/. If this function does not exist in pg, we'd better don't implement it.

In fact, Pg never has an official documentation about all functions it provided, but we can find them from pg_catalog.

```sql
select * from pg_catalog.pg_proc where proname = 'booleq';
 oid | proname | pronamespace | proowner | prolang | procost | prorows | provariadic | prosupport | prokind | prosecdef | proleakproof | proisstrict | proretset | provolatile | proparallel | pronargs | pronargdefaults | prorettype | proargtypes | proallargtypes | proargmodes | proargnames | proargdefaults | protrftypes | prosrc | probin | prosqlbody | proconfig | proacl
-----+---------+--------------+----------+---------+---------+---------+-------------+------------+---------+-----------+--------------+-------------+-----------+-------------+-------------+----------+-----------------+------------+-------------+----------------+-------------+-------------+----------------+-------------+--------+--------+------------+-----------+--------
  60 | booleq  |           11 |       10 |      12 |       1 |       0 |           0 | -          | f       | f         | t            | t           | f         | i           | s           |        2 |               0 |         16 | 16 16       |                |             |             |                |             | booleq |        |            |           |
(1 row)
```

Although materialize doesn't support it, if you think we should refer to them, I can submit an PR to them and I guess they will accept it :)

**Review by xiangjinwu — COMMENTED** [review]
What about mapping them to `Equal` and `NotEqual` after validating argument types in `bind_function`? Sth like:
```
"booleq" => {
  // some code to ensure inputs.len() == 2
  inputs[0] = inputs[0].cast_implicit()?;
  inputs[1] = inputs[1].cast_implicit()?;
  ExprType::Equal
}
```

To answer the question about these examples:
```
SELECT booleq('true','true');
SELECT 1 where 'true' is true;
SELECT 1 where 'true' and 'true';
```
In PostgreSQL, literals like `'true'` has a `unknown` type initially, and then converted to a concrete type during operator/function overload [resolution](https://www.postgresql.org/docs/current/typeconv-oper.html). This resolution process have not been implemented yet.

**Comment by neverchanje** [comment]
Good to know 😄 

> In fact, Pg never has official documentation about all functions it provided, only a subset of functions is listed in their documentation. but we can find them from pg_catalog.
> 
> ```sql
> select * from pg_catalog.pg_proc where proname = 'booleq';
>  oid | proname | pronamespace | proowner | prolang | procost | prorows | provariadic | prosupport | prokind | prosecdef | proleakproof | proisstrict | proretset | provolatile | proparallel | pronargs | pronargdefaults | prorettype | proargtypes | proallargtypes | proargmodes | proargnames | proargdefaults | protrftypes | prosrc | probin | prosqlbody | proconfig | proacl
> -----+---------+--------------+----------+---------+---------+---------+-------------+------------+---------+-----------+--------------+-------------+-----------+-------------+-------------+----------+-----------------+------------+-------------+----------------+-------------+-------------+----------------+-------------+--------+--------+------------+-----------+--------
>   60 | booleq  |           11 |       10 |      12 |       1 |       0 |           0 | -          | f       | f         | t            | t           | f         | i           | s           |        2 |               0 |         16 | 16 16       |                |             |             |                |             | booleq |        |            |           |
> (1 row)
> ```
> 
> Although materialize doesn't support it, if you think we should refer to them, I can submit a PR to them and I guess they will accept it :)

**Comment by likzn** [comment]
> What about mapping them to `Equal` and `NotEqual` after validating argument types in `bind_function`? Sth like:
> 
> ```
> "booleq" => {
>   // some code to ensure inputs.len() == 2
>   inputs[0] = inputs[0].cast_implicit(DataType:Bool)?;
>   inputs[1] = inputs[1].cast_implicit(DataType:Bool)?;
>   ExprType::Equal
> }
> ```
> 
> To answer the question about these examples:
> 
> ```
> SELECT booleq('true','true');
> SELECT 1 where 'true' is true;
> SELECT 1 where 'true' and 'true';
> ```
> 
> In PostgreSQL, literals like `'true'` has a `unknown` type initially, and then converted to a concrete type during operator/function overload [resolution](https://www.postgresql.org/docs/current/typeconv-oper.html). This resolution process have not been implemented yet.

@xiangjinwu Yes, its better. Should we define a function enum to them? Something like below, and we can have a common func to validate `len(arg)` and `cast_implicit`, then to map basic operator. And also easy to maintain these `mapping function`.
```rust
pub enum FunctionKind {
    BoolEq(true,true),
    BoolNe(true,true),
    BoolLt(true,true),
    BoolGt(true,true),
    ...
}
```

**Comment by xiangjinwu** [comment]
> Should we define a function enum to them? Something like below, and we can have a common func to validate `len(arg)` and `cast_implicit`, then to map basic operator. And also easy to maintain these `mapping function`.
> 
> ```rust
> pub enum FunctionKind {
>     BoolEq(true,true),
>     BoolNe(true,true),
>     BoolLt(true,true),
>     BoolGt(true,true),
>     ...
> }
> ```

Sorry I may not get what you meant completely. It seems to be a general and extensible framework to handle this task: map func_a to func_b with extra (stricter) argument validation.

I would prefer a lighter-weight solution than this. It is good to avoid duplication between `booleq` and `boolne`, but we should not go too far in this direction: `boollt`, `int2eq`, `int4ne`, `int8lt`, `int24gt`, `int42ge`, ...

Stepping back a little bit, we map `booleq` to `ExprType::Equal` to support it with least effort reusing existing implementation. On the backend, `ExprType::Equal` is actually dispatched back to `general_eq::<bool, bool, bool>`. In the ideal world, these 2 mappings should be avoided - if we already know both arguments are bool, it should execute `general_eq::<bool, bool, bool>`.

This is actually why PostgreSQL has these `booleq`, `int2eq` or even `int24eq` functions in addition to `=`. It does the one-way [mapping](https://github.com/postgres/postgres/blob/1d8ef62f6e1982c3019fc044dea070a8861a91bf/src/include/catalog/pg_operator.dat#L492-L496) from general `=` to a concrete `int24eq`. As it is implemented in C, it has to define these function names for all combinations rather than writing a single `general_eq::<T1, T2, T3>`. cc @TennyZhuang @neverchanje 

To summarize, if we really want to support all of these `int24eq` stuff in the future (to reach a great level of compatibility), it should map to `general_eq<int2, int4, int4>` directly, which depends on #1666. Before we have that ability, just use a lighter way to connect `xxxeq - ExprType::Equal - general_eq::<xxx>` for a small but essential subset.

**Comment by TennyZhuang** [comment]
@xiangjinwu PTAL

**Review by xiangjinwu — APPROVED** [review]
